1/2 TextInput accessibilityErrorMessage (Talkback, Android)

fabriziobertoglio1987 opened the pull request:
## Summary

**Android**: The functionality consists of calling the [AccessibilityNodeInfo#setError][10] and [#setContentInvalid][13] method to display the error message in the TextInput. 

**Fixes [#30848][51] - Adding an accessibilityErrorMessage prop to the TextInput Component**:
**Android**: The prop accessibilityErrorMessage triggers the AccessibilityNodeInfo method [setError][10] which automatically sets the correct properties on the AccessibilityNodeInfo that will inform screen readers of this state. The method calls setContentInvalid(true) and setError(youErrorString) on the AccessibilityNodeInfo.

**Fixes [#30859][52] -  Detecting changes in the Error state (text inputs)**
**Fabric - Android** - Adding accessibilityErrorMessage to field AndroidTextInputState.
ReactTextInputManager and ReactEditText receive state updates both from [Javascript][32] and [cpp (fabric)][34].
- accessibilityErrorMessage is added to the fabric AndroidTextInputState field
- The updates are received in the ReactAndroid API with method updateState from ReactTextInputManager
- After updating the TextInput text with onChangeText, the update of the accessibilityErrorMessage is triggered with method maybeSetAccessibilityError which triggers [setError][10].

More info:
- An explanation of [state updates between fabric and ReactAndroid for the TextInput component][34]
- [ReactNative renderer state updates][35]

**Paper - Android** - Adding accessibilityErrorMessage to ReactTextInputShadowNode to trigger updates in Paper renderer when accessibilityErrorMessage is changed within the onChange callback.

Related Links (Android):
- [In this diff I'm shipping and deleting mapBufferSerialization for Text measurement][101]
- [This diff implement and integrates Mapbuffer into Fabric text measure system][39]
- [Refactor ViewPropsMapBuffer -> general MapBuffer props mechanism][100]
- [TextInput: support modifying TextInputs with multiple Fragments (Cxx side)][24]
- [TextInput: keep C++ state in-sync with updated AttributedStrings in Java][23]
- [AccessibilityNodeInfo#setError][11]
- [Explanation on how TextInput calls SET_TEXT_AND_SELECTION in Java API][32]
- [Fabric: convertRawProp was extended to accept an optional default value][27]
- [understanding onChangeText callback][31]
- [Editable method replace()][12]
- [Change of error state from onChangeText show/hides a TextInput error][30]
- [AndroidTextInput: support using commands instead of setNativeProps (native change)][25]
- [TextInput: support editing completely empty TextInputs][26]
- [[Android] Fix letters duplication when using autoCapitalize #29070][40]
- [Support optional types for C++ TurboModules][28]
- [discussion on using announceForAccessibility in ReactEditText][36]
- [ fix annoucement delayed to next character][61]
- [Announce accessibility state changes happening in the background][29]
- [Refactor MountingManager into MountingManager + SurfaceMountingManager][37]

iOS Functionalities are included in separate PR https://github.com/facebook/react-native/pull/35908
Documentation PR https://github.com/facebook/react-native-website/pull/3010

Next PR [2/2 TextInput accessibilityErrorMessage (VoiceOver, iOS) #35908](https://github.com/facebook/react-native/pull/35908)
Related https://github.com/facebook/react-native-deprecated-modules/pull/18

## Changelog

[Android] [Added] - Adding TextInput prop accessibilityErrorMessage to announce with TalkBack screenreaders

## Test Plan

**Android - 20 Jan 2023**
https://github.com/facebook/react-native/pull/33468#issuecomment-1398228674

**iOS - 20 Jan 2023**
https://github.com/facebook/react-native/pull/33468#issuecomment-1398249006


<details><summary>CLICK TO OPEN OLD VIDEO TEST CASES</summary>
<p>

**PR Branch - Android and iOS 24th June**
[88]: Android - accessibilityValue announces correctly with/out errorMessage set with onChangeText or with outside event (Fabric) ([link][88])

**PR Branch - Android**
[1]. Test Cases of the functionality (Fabric) ([link][1])
[2]. Test Cases of the functionality (Paper) ([link][2])

**Main Branch**
[6]. Android - Runtime Error in main branch when passing value of 1 to TextInput  placeholder prop ([link][6])

**Issues Solved**
[7]. TalkBack error does not clear error on the next typed character when using onChangeText ([link][7])
**Other Tests**
[8]. Setting the TextInput errorMessage state with setTextAndSelection Java API from JavaScript ([link][8])
[9]. Setting the TextInput errorMessage state from fabric TextInput internal state to Java ReactTextUpdate API ([link][9])

</p>
</details>

<!-- Test Cases -->
[1]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1072101477 "Test Cases of the functionality (Android - Fabric)"
[2]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1105964322 "Test Cases of the functionality (Android - Paper)"
[3]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1116329282 "Test Cases of the functionality (iOS - Fabric)"
[6]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1087020844 "Runtime Error in main branch when passing value of 1 to TextInput  placeholder prop"
[7]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1096086753 "TalkBack error announcement done on next typed character with onChangeText"
[8]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1082594363 "setting the TextInput errorMessage state with setTextAndSelection Java API from JavaScript"
[9]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1082598745 "Setting the TextInput errorMessage state from fabric TextInput internal state to Java ReactTextUpdate API"

<!-- Links -->
[10]: https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo#setError(java.lang.CharSequence) "AOSP setError"
[11]: https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo#setError(java.lang.CharSequence) "AccessibilityNodeInfo#setError"
[12]: https://github.com/aosp-mirror/platform_frameworks_base/blob/1ac46f932ef88a8f96d652580d8105e361ffc842/core/java/android/text/Editable.java#L28-L52 "Editable method replace"
[13]: https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo#setContentInvalid(boolean) "setContentInvalid"

<!-- Commits -->
[20]: https://github.com/facebook/react-native/commit/60b6c9be8e811241039a6db5dc906a0e88e6ba82 "draft implementation of android_errorMessage "
[21]: https://github.com/facebook/react-native/commit/012d92d0b7e5de2436f186cdbff32ba128e537d5 "add errorMessage to ReactTextUpdate and maybeSetAccessibilityError"
[22]: https://github.com/fabriziobertoglio1987/react-native/commit/cad239bded5748753cee2266c27809e24c6199fb "rename android_errorMessage to errorMessageAndroid"
[23]: https://github.com/fabriziobertoglio1987/react-native/commit/0bae47434ef79eb606c453c5be8105b8df00783a "TextInput: keep C++ state in-sync with updated AttributedStrings in Java"
[24]: https://github.com/fabriziobertoglio1987/react-native/commit/0556e86d09404105dc7ff695686b8b7c01911c5c "TextInput: support modifying TextInputs with multiple Fragments (Cxx side)"
[25]: https://github.com/fabriziobertoglio1987/react-native/commit/7ab5eb4cafdea695c4c53ce2a737f6302afd6380 "AndroidTextInput: support using commands instead of setNativeProps (native change)"
[26]: https://github.com/fabriziobertoglio1987/react-native/commit/b9491b7c5104066b2714045cd7710f995458c9e9 "TextInput: support editing completely empty TextInputs"
[27]: https://github.com/fabriziobertoglio1987/react-native/commit/7f1ed6848f89bdccc7f7a5cc76019eec67e76b2f "Fabric: convertRawProp was extended to accept an optional default value"
[28]: https://github.com/facebook/react-native/commit/6e0fa5f15eef71abcfb47750eb3669065ba2ab7d "Support optional types for C++ TurboModules"
[29]: https://github.com/fabriziobertoglio1987/react-native/commit/baa66f63d8af2b772dea8ff8eda50eba264c3faf "Announce accessibility state changes happening in the background"

<!-- Comments -->
[30]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1071989570 "Change of error state from onChangeText show/hides a TextInput error"
[31]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1074827746 "understanding onChangeText callback"
[32]: https://github.com/facebook/react-native/issues/29063#issuecomment-658189938 "Explanation on how TextInput calls SET_TEXT_AND_SELECTION in Java API"
[33]: https://github.com/facebook/react-native/pull/33468#discussion_r835036889 "Explanation of TextInput state management with fabric C++ and JAVA API"
[34]: https://github.com/facebook/react-native/pull/33468#discussion_r835036889 "state updates between fabric and ReactAndroid for the TextInput component"
[35]: https://reactnative.dev/architecture/render-pipeline#react-native-renderer-state-updates "ReactNative renderer state updates"
[35]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1080144483 "Analysis on how AndroidTextInputState.cpp sends updates to ReactTextInputManager"
[36]: https://github.com/facebook/react-native/pull/33468#discussion_r848162849 "discussion on using announceForAccessibility in ReactEditText"
[37]: https://github.com/fabriziobertoglio1987/react-native/commit/29eb632f1cb2ef5459253783eac43e5d7e999742 "Refactor MountingManager into MountingManager + SurfaceMountingManager"
[38]: https://github.com/fabriziobertoglio1987/react-native/commit/733f2285067de401b925195266f4cec84c3f7fef "Diff C++ props for Android consumption"
[39]: https://github.com/fabriziobertoglio1987/react-native/commit/91b3f5d48aa1322046b8c5335f8e2e1a5e702b67 "This diff implement and integrates Mapbuffer into Fabric text measure system"

<!-- PRs -->
[40]: https://github.com/facebook/react-native/pull/29070 "[Android] Fix letters duplication when using autoCapitalize #29070"

<!-- Issues -->
[50]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12  "Notes from work on iOS/Android: Text input error for screenreaders #12"
[51]: https://github.com/facebook/react-native/issues/30848 "iOS/Android: Text input error for screenreaders #30848"
[52]: https://github.com/facebook/react-native/issues/30859 "Android: Error state change (text inputs) #30859"

<!-- Commit -->
[61]: https://github.com/facebook/react-native/pull/33468/commits/eb33c933c8bcb9a8421a6acdb7a51f261121be45 "fix annoucement delayed to next character"

<!-- More tests -->
[70]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1116966512 "iOS - Paper renderer does not update the accessibilityValue"
[71]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1124631221 "Test Cases of the functionality (Fabric) after removing changes to .cpp libs"
[72]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1124892802 "Test Cases of the functionality (Paper) after removing changes to .cpp libs"
[73]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1132830758 "iOS - announcing error onChangeText and screenreader focus"
[74]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1150657065 "iOS - The screenreader announces the TextInput value after the errorMessage is cleared"
[75]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1152285978 "iOS - Exception thrown while executing UI block: - [RCTTextView setOnAccessibiltyAction:]: unrecognized selector sent to instance (Paper) (main branch)"
[76]: https://github.com/facebook/react-native/issues/30859#issuecomment-1158790381 "iOS - announce lastChar (not entire text) onChangeText and avoid multiple announcements (Fabric)"
[77]: https://github.com/facebook/react-native/issues/30859#issuecomment-1158794863 "iOS - announces or does not announce the accessibilityError through Button onPress (not onChangeText) (Fabric)"
[78]: https://github.com/facebook/react-native/issues/30859#issuecomment-1158797801 "iOS - the error is announced with accessibilityInvalid true and does not clear after typing text (onChangeText) (Fabric)"
[79]: https://github.com/facebook/react-native/issues/30848#issuecomment-1162799299 "iOS - Exception thrown while executing UI block: - RCTUITextView setAccessibilityErrorMessage:]: unrecognized selector sent to instance (iOS - Paper on main branch)"

<!-- Links iOS -->
[80]: https://github.com/fabriziobertoglio1987/react-native/commit/e13b9c6e49480e8262df06b7c1e99caab74e801f "RCTTextField was spliited into two classes"
[81]: https://github.com/fabriziobertoglio1987/react-native/commit/ee9697e5155aa972564d5aac90ceeb9db100750d "Introducing RCTBackedTextInputDelegate"
[82]: https://github.com/fabriziobertoglio1987/react-native/commit/2dd2529b3ab3ace39136a6e24c09f80ae421a17e "Add option to hide context menu for TextInput"
[83]: https://github.com/fabriziobertoglio1987/react-native/blob/343eea1e3150cf54d6f7727cd01d13eb7247c7f7/React/Fabric/Mounting/ComponentViews/Text/RCTParagraphComponentAccessibilityProvider.mm#L48-L72 "RCTParagraphComponentAccessibilityProvider accessibilityElements"
[84]: https://github.com/fabriziobertoglio1987/react-native/blob/c8790a114f6f21774c43f0e9b9210e7b35d1c243/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm#L613 "RCTTextInputComponentView method _setAttributedString"
[85]: https://github.com/fabriziobertoglio1987/react-native/blob/c8790a114f6f21774c43f0e9b9210e7b35d1c243/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm#L146 "RCTTextInputComponentView method updateProps"
[86]: https://github.com/fabriziobertoglio1987/react-native/blob/c8790a114f6f21774c43f0e9b9210e7b35d1c243/Libraries/Text/TextInput/RCTBaseTextInputView.m#L150 "RCTBaseTextInputView setAttributedText"
[87]: https://github.com/facebook/react-native/issues/30859#issuecomment-1165395361 "iOS - accessibilityValue announces correctly with/out errorMessage set with onChangeText or with outside event"
[88]: https://github.com/facebook/react-native/issues/30859#issuecomment-1165398153 "Android - accessibilityValue announces correctly with/out errorMessage set with onChangeText or with outside event"
[89]: https://github.com/facebook/react-native/issues/30859#issuecomment-1165413245 "iOS - accessibilityValue announces correctly with/out errorMessage set with onChangeText or with outside event (Fabric)"

<!-- Links Android -->
[100]: https://github.com/fabriziobertoglio1987/react-native/commit/110b191b14e3cb692bb6a33f0f129b4f0215f9a6 "Refactor ViewPropsMapBuffer -> general MapBuffer props mechanism"
[101]: https://github.com/fabriziobertoglio1987/react-native/commit/22b6e1c8ec0e69700e9142cf5c9c1ab1e6a84b78 "In this diff I'm shipping and deleting mapBufferSerialization for Text measurement"

fabriziobertoglio1987 commented on ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java:
```diff
@@ -1320,12 +1326,17 @@ public Object updateState(
         TextAttributeProps.getTextBreakStrategy(paragraphAttributes.getString("textBreakStrategy"));
 
+    @Nullable
+    String errorMessage =
+        props.hasKey("errorMessageAndroid") ? props.getString("errorMessageAndroid") : null;
+
     return ReactTextUpdate.buildReactTextUpdateFromState(
         spanned,
         state.getInt("mostRecentEventCount"),
         TextAttributeProps.getTextAlignment(props, TextLayoutManager.isRTL(attributedString)),
         textBreakStrategy,
         TextAttributeProps.getJustificationMode(props),
-        containsMultipleFragments);
+        containsMultipleFragments,
+        errorMessage);
```
This code was introduced to fix the issue https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1071989570. 

<details><summary>video test without this solution</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/158931170-38da40a4-8f0e-4a49-bed4-8ed6f84221c4.mp4" width="1000" />

</p>
</details>


<details><summary>video test with this solution</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/158931196-21426511-427c-485f-8975-040e18e557c5.mp4" width="1000" />

</p>
</details>

I reviewed again this solution and found the fabric code responsible for triggering the ReactTextUpdate

The text and paragraphAttributes state is passed to Java from C++

https://github.com/facebook/react-native/blob/05b0d29e765be1f60532bd5e5243f193e8828ddb/ReactCommon/react/renderer/components/textinput/androidtextinput/react/renderer/components/androidtextinput/AndroidTextInputState.cpp#L85-L87

The Java ReactAndroid API receives the state update with the method updateState

https://github.com/facebook/react-native/blob/6780e00fcb515af04ff91fba7f28ff7152e3210f/ReactAndroid/src/main/java/com/facebook/react/uimanager/StateWrapper.java#L15-L20

https://github.com/facebook/react-native/blob/6780e00fcb515af04ff91fba7f28ff7152e3210f/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java#L1283-L1285

My current solution is retrieving the errorMessage from the `props` and adding it to the call to `ReactTextUpdate.buildReactTextUpdateFromState` as 8th parameter, but I believe this is not the best solution:
- other props (for ex. `underlineColorAndroid`) include some c++ mappings in the androidtextinput/*.cpp files
- I believe similar c++ configs to `underlineColorAndroid` should be done for errorMessageAndroid 
- `errorMessageAndroid` should trigger a change in state from the above c++ API

More info are included in https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1078727833 and https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1078692146

fabriziobertoglio1987 replied:
commit https://github.com/facebook/react-native/pull/33468/commits/aca0d9989fd4e403aeec6082c1df542836372e8e

fabriziobertoglio1987 reviewed (commented):
@blavalla I decided to not move this PR to Ready for Review, even if is a [perfectly working functionality][2]. 

I'm trying to add the changes explained below to fabric androidtextinput and related libraries.
This solution works and does not introduce any regressions, but I think that I can improve it. 
I would like to work on these improvements at the beginning of next week, but If you believe they are not necessary. I would be happy to move the PR to Ready for Review. I included this information in the PR Summary and they will be part of the commit.
Below a detailed explanation. Thanks a lot. I remain available also during the weekend if required. :pray: 

[1]: https://github.com/facebook/react-native/blob/05b0d29e765be1f60532bd5e5243f193e8828ddb/ReactCommon/react/renderer/components/textinput/androidtextinput/react/renderer/components/androidtextinput/AndroidTextInputState.cpp#L85-L87 "AndroidTextInputState"
[2]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1072101477 "test cases of functionality"

fabriziobertoglio1987 commented:
The documentation is included in PR https://github.com/facebook/react-native-website/pull/3010

blavalla commented on ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java:
```diff
@@ -504,6 +518,22 @@ public int incrementAndGetEventCounter() {
+    if (!canUpdateWithEventCount(eventCounter) || accessibilityErrorMessage == null) {
+      return;
+    }
+    announceForAccessibility("Invalid input " + accessibilityErrorMessage);
```
What is the use case for this? As far as I know, Talkback should automatically announce error messages when they are set on the view, assuming we are firing the proper "content change" event that Talkback picks up here - https://github.com/google/talkback/blob/master/compositor/src/main/res/raw/compositor.json#L380-L387

Was this fixing a specific case?

fabriziobertoglio1987 replied:
Thanks @blavalla. This solution had to be implemented to fix the following issue:

<details><summary>when setting accessibilityErrorAndroid onChangeText, TalkBack announces the error on the next typed character</summary>
<p>

- The user types the word `erro` in the TextInput field. 
- The user then types the last character `r` , the JavaScript component triggers the error with the prop `accessibilityErrorAndroid`

**Expected Result**:
when the user types error, the screenreader announces an error

**Actual Result**:
the screenreader does not announce the error when the text is equal to error, but on the next typed/deleted characters (errore) and (erro). There is a delay of 1 character between passing the error to the TextInput and TalkBack announcing it.

<details><summary>click to open code used in the test</summary>
<p>

```javascript
function ErrorExample(): React.Node {
  const [text, setText] = React.useState('');
  const [error, setError] = React.useState(null);
  return (
    <>
      <Button onPress={() => setError('button')} title="Press to set error" />
      <Text>
        Type error in the below TextInput to display an error message.
      </Text>
      <TextInput
        screenreaderErrorAndroid={error}
        onBlur={() => setError('onBlur')}
        onEndEditing={() => setError('onEndEditing')}
        onChangeText={newText => {
          setText(newText);
          if (newText === 'error') {
            setError('the newText is: ' + newText);
          } else if (newText === 'empty') {
            setError('');
          } else if (newText === 'null') {
            setError(null);
          } else if (error !== 'onBlur') {
            setError(null);
          }
        }}
        value={text}
        style={styles.default}
      />
    </>
  );
}
```
</p>
</details>

<details><summary>video showcasing the problem</summary>
<p>

<video src="https://user-images.githubusercontent.com/24992535/162851364-88c02366-0bbf-4174-bf5b-359c3247ab20.mp4" width="1000" />

</p>
</details>

</p>
</details>

The test case [TalkBack error announcement done on next typed character with onChangeText][11] investigates the issue in more detail:

The method maybeSetAccessibilityError is called from [fabric state updates][1] after the TalkBack accessibility announcement ([screenshot 1][3], [screenshot 2][4], [screenshot 3][5]).
The solution for this issue is triggering manually the accessibility announcement from `maybeSetAccessibilityError` and clearing  R.id.accessibility_error before the user types the next letter.

- `announceForAccessibility` triggers the accessibility error onChangeText
- `setTag(R.id.accessibility_error, null)`  clear the error before the next character is typed, otherwise the error is announced also on the next/previous character

```java
public void maybeSetAccessibilityError(
    int eventCounter, @Nullable String accessibilityErrorMessage) {
  if (!canUpdateWithEventCount(eventCounter) || accessibilityErrorMessage == null) {
    return;
  }
  announceForAccessibility("Invalid input " + accessibilityErrorMessage);
  setTag(R.id.accessibility_error, null);
}
```

[1]: https://github.com/fabriziobertoglio1987/react-native/blob/994089cfbe106ac839a07b90cb4e851a3f0aa7e3/ReactCommon/react/renderer/components/textinput/androidtextinput/react/renderer/components/androidtextinput/AndroidTextInputShadowNode.cpp#L118 "fabric updateStateIfNeeded" 
[2]: https://user-images.githubusercontent.com/24992535/162907581-c03239e7-233a-4e24-b4e6-fbf493a69040.png "state update"
[3]: https://user-images.githubusercontent.com/24992535/162907581-c03239e7-233a-4e24-b4e6-fbf493a69040.png "screenshot -  the state sent from fabric"
[4]: https://user-images.githubusercontent.com/24992535/162907091-f2e779fe-91ea-45c8-b83f-ccc50af0a361.png "screenshot - the ScreenReader announcement is done before receiving state update from fabric"
[5]: https://user-images.githubusercontent.com/24992535/162903238-48a7eb45-74ff-460a-b9a8-29256347a2b1.png "screenshot - after receiving the state update from fabric"

[10]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1095709282 "when setting error onChangeText, announcement is done on the next typed character"
[11]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1096086753 "TalkBack error announcement done on next typed character with onChange"

fabriziobertoglio1987 replied:
Thanks, @blavalla. The new solution I implemented is:

1) The method `maybeSetAccessibilityError` receives updates from fabric (c++).
     - The method is called when the prop screenreaderErrorAndroid is updated in the TextInput onChangeText callback
     - The method triggers the event `TYPE_WINDOW_CONTENT_CHANGED`

https://github.com/facebook/react-native/blob/0513007362b419c8762a0ba9b954c231d6cb595b/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java#L546-L555

2) `requestSendAccessibilityEvent` is invoked to fix the Issue [TalkBack error announcement done on next typed character with onChangeText][1], avoiding to announce the error message twice (on text "error" and next character "errore").

[1]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1096086753

https://github.com/facebook/react-native/blob/0513007362b419c8762a0ba9b954c231d6cb595b/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java#L176-L181

The solution is taken from the [TextView#setError][2] method which calls `notifyViewAccessibilityStateChangedIfNeeded`. I remain available to do further improvements. :pray: 

[2]: https://github.com/aosp-mirror/platform_frameworks_base/blob/1ac46f932ef88a8f96d652580d8105e361ffc842/core/java/android/widget/TextView.java#L7284-L7285

fabriziobertoglio1987 replied:
Thanks @blavalla, I updated the videos of the examples below.

## Test Plan

**PR Branch**
[1]. Test Cases of the functionality (Fabric) ([link][1])
[2]. Test Cases of the functionality (Paper) ([link][2])

**Issues Solved**
[6]. TalkBack error does not clear error on the next typed character when using onChangeText ([link][6])

<!-- Test Cases -->
[1]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1072101477 "Test Cases of the functionality (Fabric)"
[2]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1105964322 "Test Cases of the functionality (Paper)"
[4]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1087020844 "Runtime Error in main branch when passing value of 1 to TextInput  placeholder prop"
[5]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1095709282 "when setting accessibilityErrorAndroid onChangeText, TalkBack announces the error on the next typed character"
[6]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1096086753 "TalkBack error announcement done on next typed character with onChangeText"
[7]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1082594363 "setting the TextInput errorMessage state with setTextAndSelection Java API from JavaScript"
[8]: https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1082598745 "Setting the TextInput errorMessage state from fabric TextInput internal state to Java ReactTextUpdate API"

fabriziobertoglio1987 commented:
Thanks, @kacieb. I'm currently working on iOS side of this functionality and I plan to complete it by Friday 6th of May.

fabriziobertoglio1987 commented:
Thanks @blavalla. Once I complete the changes for code review [Android: Role description is announced before the components text, rather than after #31042](https://github.com/facebook/react-native/issues/31042), I will further test the iOS functionality of the Issue [Text input error for screenreaders #30848](https://github.com/facebook/react-native/pull/33468) and fix any issue in the iOS functionalities or think about potential improvements to the PR on the iOS side.

The PR summary includes a test case for iOS [`iOS - announcing error onChangeText and screenreader focus`](https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1132830758).

fabriziobertoglio1987 reviewed (commented):
The below commits gave me some ideas on how to improve the iOS solution in the future. I may invest additional time to work on this in the future, following the patterns used in previous commits.
- [Diff C++ props for Android consumption][38]
- [RCTTextField was spliited into two classes][80]
- [Introducing RCTBackedTextInputDelegate][81]
- [Add option to hide context menu for TextInput][82]
- [RCTParagraphComponentAccessibilityProvider accessibilityElements][83]

[38]: https://github.com/fabriziobertoglio1987/react-native/commit/733f2285067de401b925195266f4cec84c3f7fef "Diff C++ props for Android consumption"
[80]: https://github.com/fabriziobertoglio1987/react-native/commit/e13b9c6e49480e8262df06b7c1e99caab74e801f "RCTTextField was spliited into two classes"
[81]: https://github.com/fabriziobertoglio1987/react-native/commit/ee9697e5155aa972564d5aac90ceeb9db100750d "Introducing RCTBackedTextInputDelegate"
[82]: https://github.com/fabriziobertoglio1987/react-native/commit/2dd2529b3ab3ace39136a6e24c09f80ae421a17e "Add option to hide context menu for TextInput"
[83]: https://github.com/fabriziobertoglio1987/react-native/blob/343eea1e3150cf54d6f7727cd01d13eb7247c7f7/React/Fabric/Mounting/ComponentViews/Text/RCTParagraphComponentAccessibilityProvider.mm#L48-L72 "RCTParagraphComponentAccessibilityProvider accessibilityElements"

necolas commented:
@kacieb equivalent functionality in terms of web aria api https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-errormessage

kacieb commented:
> @kacieb equivalent functionality in terms of web aria api https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-errormessage

This is awesome @fabriziobertoglio1987! Would it be possible to follow the aria names and pattern for the JavaScript API? No native changes would be needed, you can do all of this in JavaScript I think. So instead of:

> The acceptable parameters are null or a string. The parameter null removes the error message.

It would be something like

> The acceptable parameters of `accessibilityErrorMessage` is a string. Setting `accessibilityInvalid` to `true` activates the error message. Setting `accessibilityInvalid` to `false` removes the error message.

Instead of checking for `null` on the error message to clear it, you'd check the `accessibilityInvalid` boolean for when to set or clear the message.

Does that make sense? Please let me know if I can help explain better!

cc @necolas

necolas commented:
something to note is that the value of aria-errormessage is meant to be the id of an element that contains the error message.

kacieb commented:
> something to note is that the value of aria-errormessage is meant to be the id of an element that contains the error message.

Good note. I think this is out of scope for this PR. Definitely something we can improve in the future.

fabriziobertoglio1987 reviewed (commented):
- [x] Implement solution for multiline TextInput (iOS)
- [ ] Further review and test the pull request

fabriziobertoglio1987 commented on Libraries/Text/TextInput/RCTBaseTextInputView.m:
```diff
@@ -145,12 +145,20 @@ - (void)setAttributedText:(NSAttributedString *)attributedText
+  if (currentScreenreaderError == nil && errorMessageRemoved) {
+    BOOL validString = attributedText && [attributedText.string length] != 0;
+    NSString *lastChar = validString ? [attributedText.string substringFromIndex:[attributedText.string length] - 1] : @"";
+    self.backedTextInputView.accessibilityValue = attributedText.string;
```
setting accessibilityValue to nil is enought to fix this regression. This logic has to be removed

https://developer.apple.com/documentation/uikit/uiaccessibilityelement/1619583-accessibilityvalue?language=objc

fabriziobertoglio1987 replied:
https://github.com/fabriziobertoglio1987/react-native/commit/16d7d4d8494a97f162ad8429043f6b5f334fd019

fabriziobertoglio1987 reviewed (commented):
- [x] test iOS Paper and Fabric
- [x] test Android Paper and Fabric
- [x] try to reproduce [runtime](https://github.com/fabriziobertoglio1987/react-native-notes/issues/12#issuecomment-1153457291) on main branch
- [x] review remaining diff

fabriziobertoglio1987 commented on Libraries/Text/TextInput/Multiline/RCTUITextView.m:
```diff
@@ -120,6 +120,22 @@ - (void)setDefaultTextAttributes:(NSDictionary<NSAttributedStringKey, id> *)defa
+{
+  self.previousScreenreaderError = self.currentScreenreaderError;
+  self.currentScreenreaderError = accessibilityErrorMessage;
+  NSString *text = self.attributedText.string;
```
https://github.com/fabriziobertoglio1987/react-native/commit/16d7d4d8494a97f162ad8429043f6b5f334fd019

fabriziobertoglio1987 commented on Libraries/Text/TextInput/RCTBaseTextInputView.m:
```diff
@@ -145,12 +145,20 @@ - (void)setAttributedText:(NSAttributedString *)attributedText
+  NSString *currentScreenreaderError = self.backedTextInputView.currentScreenreaderError;
+  NSString *previousScreenreaderError = self.backedTextInputView.previousScreenreaderError;
+  BOOL errorMessageRemoved = currentScreenreaderError == nil && ![currentScreenreaderError isEqualToString: previousScreenreaderError];
+  if (currentScreenreaderError == nil && errorMessageRemoved) {
```
https://github.com/fabriziobertoglio1987/react-native/commit/16d7d4d8494a97f162ad8429043f6b5f334fd019

fabriziobertoglio1987 commented on Libraries/Text/TextInput/RCTBaseTextInputView.m:
```diff
@@ -145,12 +145,19 @@ - (void)setAttributedText:(NSAttributedString *)attributedText
 
+  NSString *currentScreenreaderError = self.backedTextInputView.currentScreenreaderError;
+  NSString *previousScreenreaderError = self.backedTextInputView.previousScreenreaderError;
+  BOOL errorMessageRemoved = currentErrorMessageString == nil && currentErrorMessageString != previousErrorMessageString;
```
https://github.com/fabriziobertoglio1987/react-native/commit/16d7d4d8494a97f162ad8429043f6b5f334fd019

fabriziobertoglio1987 commented on React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:
```diff
@@ -134,6 +135,22 @@ - (void)updateProps:(Props::Shared const &)props oldProps:(Props::Shared const &
+      UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, errorWithLastCharacter);
+    } else {
+      self.accessibilityElement.accessibilityValue = text;
+      self->_errorMessage = nil;
```
- [ ] rename prop
- [ ] apply same solution from Paper (low priority)

fabriziobertoglio1987 replied:
https://github.com/facebook/react-native/pull/33468/commits/933aa3cd074865a6ab402312de0515b6840ba290

fabriziobertoglio1987 commented on React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:
```diff
@@ -585,6 +602,14 @@ - (void)_setAttributedString:(NSAttributedString *)attributedString
   UITextRange *selectedRange = _backedTextInputView.selectedTextRange;
   NSInteger oldTextLength = _backedTextInputView.attributedText.string.length;
   _backedTextInputView.attributedText = attributedString;
+  if (self->_errorMessage == nil) {
```
https://github.com/fabriziobertoglio1987/react-native/commit/30f3f963da4c666e40b5e806ce86eb625661417d

fabriziobertoglio1987 commented on Libraries/Text/TextInput/Singleline/RCTUITextField.m:
```diff
@@ -98,6 +98,22 @@ - (void)setEditable:(BOOL)editable
 
+- (void)setAccessibilityErrorMessage:(NSString *)accessibilityErrorMessage
+{
+  self.previousScreenreaderError = self.currentScreenreaderError;
```
use mutableCopy https://www.techotopia.com/index.php/Copying_Objects_in_Objective-C

fabriziobertoglio1987 replied:
https://github.com/fabriziobertoglio1987/react-native/commit/ce5e52ad843396d60b977bb770f54b31477d31fa

fabriziobertoglio1987 commented on React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:
```diff
@@ -134,6 +136,23 @@ - (void)updateProps:(Props::Shared const &)props oldProps:(Props::Shared const &
   }
 
+  if (newTextInputProps.accessibilityErrorMessage != oldTextInputProps.accessibilityErrorMessage) {
+    self->previousScreenreaderError =  self->currentScreenreaderError;
```
https://github.com/fabriziobertoglio1987/react-native/commit/ce5e52ad843396d60b977bb770f54b31477d31fa

fabriziobertoglio1987 commented:
- [ ] fix regression in Fabric iOS. Reintroduce solution from commit https://github.com/facebook/react-native/pull/33468/commits/318a96420f0593dc59c030193009e9cba31eb666
- [ ] extensive iOS and Android tests
- [ ] review diff and PR summary

kacieb commented:
This looks good to me on the JS side! I'll work with blavalla to make sure we can land safely. cc @lunaleaps for awareness.

blavalla reviewed (approved):
Looks good from an Android/iOS perspective.

The "lastChar" bit on iOS is a bit strange, but it does properly replicate iOS's keyboard echoing behavior with VoiceOver. I'll add a comment around this when I merge it in, as I think it makes sense to document why it's there.

lunaleaps commented:
@fabriziobertoglio1987 Do we have accompanying API documentation on the website for this?

yungsters commented on Libraries/Components/TextInput/AndroidTextInputNativeComponent.js:
```diff
@@ -165,6 +165,19 @@ export type NativeProps = $ReadOnly<{|
     'off',
   >,
 
+  /**
+   * String to be read by screenreaders to indicate an error state. The acceptable parameters
+   * of accessibilityErrorMessage is a string. Setting accessibilityInvalid to true activates
+   * the error message. Setting accessibilityInvalid to false removes the error message.
+   */
+  accessibilityErrorMessage?: ?Stringish,
+
+  /**
+   * Setting accessibilityInvalid to true activates the error message. Setting accessibilityInvalid
+   * to false removes the error message.
+   */
+  accessibilityInvalid?: ?boolean,
+
```
What is the rationale for adding two props when they could instead be one? (Also, probably no need to reiterate in the type of the prop in the docblock, since the type is already defined and will be presumably extracted for any reference documentation purposes.)

```suggestion
  /**
   * String to be read by screen readers to indicate an error state. Setting this to null or
   * undefined will remove the error message.
   */
  accessibilityErrorMessage?: ?Stringish,
```

fabriziobertoglio1987 replied:
>What is the rationale for adding two props when they could instead be one? 

It was introduced with the code review https://github.com/facebook/react-native/pull/33468#issuecomment-1147723451 

Adapts this API to the Web API aria-errormessage
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-errormessage

>When there is a user-created error, you want to let them know it exists and tell them how to fix it. There are two attributes you need to use: set [aria-invalid="true"](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-invalid) to define the object as being in an error state, then add the aria-errormessage attribute with the value being the id of the element containing the error message text for that object.

This API (`accessibilityErrorMessage` and `accessibilityInvalid`) is based on Android [AccessbilityNodeInfo](https://github.com/aosp-mirror/platform_frameworks_base/blob/00545d0fd6ba561271b690bac51f3439d3bc56e5/core/java/android/view/accessibility/AccessibilityNodeInfo.java#L3005-L3008) API. The API allows adding accessibility information to each View node and announcing it with [talkback](https://github.com/google/talkback).

Each android view and widget (EditText, TextView or View) stores their accessibility information in their AccessibilityNodeInfo node.
TalkBack reads this information when the screenreader's focus changes.

The error message of an EditText is set on the AccessibillityNodeInfo using the following APIs:

- [AccessbilityNodeInfo#setError](https://github.com/aosp-mirror/platform_frameworks_base/blob/00545d0fd6ba561271b690bac51f3439d3bc56e5/core/java/android/view/accessibility/AccessibilityNodeInfo.java#L3005-L3008) to set the error text of the node.
- [AccessibilityNodeInfo#setContentInvalid](https://github.com/aosp-mirror/platform_frameworks_base/blob/00545d0fd6ba561271b690bac51f3439d3bc56e5/core/java/android/view/accessibility/AccessibilityNodeInfo.java#L2502-L2504) to set the content as invalid.  For example, a date is not well-formed.

https://github.com/facebook/react-native/pull/33468#issuecomment-1147732785
>something to note is that the value of aria-errormessage is meant to be the id of an element that contains the error message.

This API relies on the [AccessbilityNodeInfo#setError](https://github.com/aosp-mirror/platform_frameworks_base/blob/00545d0fd6ba561271b690bac51f3439d3bc56e5/core/java/android/view/accessibility/AccessibilityNodeInfo.java#L3005-L3008). It does not add a hidden text element to the View hierarchy using react-native. Instead, we use the Native Android API (AccessibilityNodeInfo) to add the error message to the node of the EditText. TalkBack announces the error message when focusing on the EditText (TextInput).

I'll be happy to remove the accessibililtyInvalid prop and further improve the PR. Thanks

fabriziobertoglio1987 replied:
> no need to reiterate in the type of the prop in the docblock, since the type is already defined and will be presumably extracted for any reference documentation purposes.

I removed the documentation in the docblock of prop accessibilityInvalid with commit https://github.com/facebook/react-native/pull/33468/commits/cb0c5060cb721974f7d39bf3fb2fa2cecb2bc7f1

lunaleaps replied:
Hmm this is interesting because it doesn't completely follow the spec since `accessibilityErrorMessage` doesn't provide the elementId but it also introduces this extra logic around `accessibilityInvalid` which is part of the web standards. 

I think I'd lean towards keeping the two separate props but including in the documentation that we're modelling this after `aria` standards. See my comment here: https://github.com/facebook/react-native-website/pull/3010#discussion_r957610518

fabriziobertoglio1987 replied:
Thanks a lot, @lunaleaps. I updated the documentation as requested with commit  https://github.com/facebook/react-native-website/pull/3010/commits/8510e8cba503d523f18053e7e2c3b3125f11c3b6  🙏 .

<details><summary>preview of the documentation</summary>
<p>

![Screen Shot 2022-08-30 at 21 03 45](https://user-images.githubusercontent.com/24992535/187443887-a4cda43d-cfc0-452e-8ad4-ecec6008b678.png)

</p>
</details>

yungsters commented:
Thanks for this great feature!

Related… when this gets landed, it would be great to also submit a PR to [`@types/react-native`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-native/index.d.ts#L1299) to add the new prop(s).

fabriziobertoglio1987 commented:
Thanks. I agree @blavalla https://github.com/facebook/react-native/pull/33468#pullrequestreview-1061123093. I added some comments and links with this commit https://github.com/facebook/react-native/pull/33468/commits/d85c90ad3fd80a408343cd3924e0af1ef0963713.

fabriziobertoglio1987 commented:
> @fabriziobertoglio1987 Do we have accompanying API documentation on the website for this?

@lunaleaps This is the PR https://github.com/facebook/react-native-website/pull/3010 for the documentation. Thanks

fabriziobertoglio1987 commented:
> Related… when this gets landed, it would be great to also submit a PR to [`@types/react-native`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-native/index.d.ts#L1299) to add the new prop(s).

I added this task to my list of TODOs, and I will prepare the PR once merged. I continue working on the next tasks (https://github.com/facebook/react-native/pull/33690#discussion_r865138542 and https://github.com/facebook/react-native/pull/34353#issuecomment-1207744977). Thanks

fabriziobertoglio1987 commented:
<details><summary>App crashing after trying to parse accessibilityErrorMessage prop raw value (rebase 14th Oct.)</summary>
<p>

The issue was solved after rebasing again on 17th Oct, including this for documentation purpose.

The app crashed at [RawProps.cpp:92](https://github.com/fabriziobertoglio1987/react-native/blob/a92e7f621ac8da36299ce6c7a82f2cbe11fd9f9a/ReactCommon/react/renderer/core/RawProps.cpp#L90-L92) with error:

```
function at: assertion failed (parser_ && "The object is not parsed. `parse` must be called before `at`.")' 
```

https://www.icloud.com/iclouddrive/0e4zx_DEQrjC6s2uDVLTLXYPQ#null_pointer_dereference

The issue would not reproduce after removing the changes to the [attributedstring](https://github.com/fabriziobertoglio1987/react-native/tree/a92e7f621ac8da36299ce6c7a82f2cbe11fd9f9a/ReactCommon/react/renderer/attributedstring) ParagraphAttributes field accessibilityErrorMessage, but the error message is not displayed onChangeText. 

The error initially seems triggered by conversions fromRawValue, RawProps, convertRawProp. The error could be caused by a conversion issue in ParagraphAttributes[accessibilityErrorMessage] and an error in parsing the value.

The fix consisted in:
- rebase again (diff from 14th Oct at 09:00 and 17th Oct 13:22 Rome)
- clean build (which is not the cause of the issue as I tried it several times)

<details><summary>crash with log</summary>
<p>

https://user-images.githubusercontent.com/24992535/196162666-62c12b88-2ba1-4f15-bcc8-31918ecc3e5c.mp4

</p>
</details>

<details><summary>after rebasing and clean build</summary>
<p>

https://user-images.githubusercontent.com/24992535/196162657-8dda9513-aa96-4de4-b661-97d9c126f39a.mp4

</p>
</details>

</p>
</details>

fabriziobertoglio1987 commented:
<details><summary>App crashing after trying to parse accessibilityErrorMessage prop raw value (rebase 14th Oct.)</summary>
<p>

The issue was solved after rebasing again on 17th Oct, including this for documentation purpose.

The app crashed at [RawProps.cpp:92](https://github.com/fabriziobertoglio1987/react-native/blob/a92e7f621ac8da36299ce6c7a82f2cbe11fd9f9a/ReactCommon/react/renderer/core/RawProps.cpp#L90-L92) with error:

```
function at: assertion failed (parser_ && "The object is not parsed. `parse` must be called before `at`.")' 
```

https://www.icloud.com/iclouddrive/0e4zx_DEQrjC6s2uDVLTLXYPQ#null_pointer_dereference

The issue would not reproduce after removing the changes to the [attributedstring](https://github.com/fabriziobertoglio1987/react-native/tree/a92e7f621ac8da36299ce6c7a82f2cbe11fd9f9a/ReactCommon/react/renderer/attributedstring) ParagraphAttributes field accessibilityErrorMessage, but the error message is not displayed onChangeText. 

The error initially seems triggered by conversions fromRawValue, RawProps, convertRawProp. The error could be caused by a conversion issue in ParagraphAttributes[accessibilityErrorMessage] and an error in parsing the value.

The fix consisted in:
- rebase again (diff from 14th Oct at 09:00 and 17th Oct 13:22 Rome)
- clean build (which is not the cause of the issue as I tried it several times)

<details><summary>crash with log</summary>
<p>

https://user-images.githubusercontent.com/24992535/196162666-62c12b88-2ba1-4f15-bcc8-31918ecc3e5c.mp4

</p>
</details>

<details><summary>after rebasing and clean build</summary>
<p>

https://user-images.githubusercontent.com/24992535/196162657-8dda9513-aa96-4de4-b661-97d9c126f39a.mp4

</p>
</details>

</p>
</details>

fabriziobertoglio1987 commented:
- [ ] add fabric .cpp configurations https://github.com/fabriziobertoglio1987/react-native/commit/110b191b14e3cb692bb6a33f0f129b4f0215f9a6

lunaleaps commented:
@fabriziobertoglio1987 Is this ready to be reviewed again?

fabriziobertoglio1987 commented on ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java:
```diff
@@ -157,6 +160,36 @@ public ReactEditText(Context context) {
+            if (event.getEventType() == AccessibilityEvent.TYPE_VIEW_TEXT_CHANGED
+                && host.getParent() != null) {
+              try {
+                host.getParent().requestSendAccessibilityEvent(host, event);
```
try to remove this and test if there is a regression on Android

fabriziobertoglio1987 commented on ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java:
```diff
@@ -539,6 +572,25 @@ public int incrementAndGetEventCounter() {
+   * @param eventCounter
+   * @param accessibilityErrorMessage
+   */
+  public void maybeSetAccessibilityError(
+      int eventCounter, @Nullable String accessibilityErrorMessage) {
+    String previousScreenreaderError = (String) getTag(R.id.accessibility_error_message);
+    if (!canUpdateWithEventCount(eventCounter)
+        || previousScreenreaderError == accessibilityErrorMessage) {
+      return;
+    }
+
+    setTag(R.id.accessibility_error_message, accessibilityErrorMessage);
+    sendAccessibilityEvent(AccessibilityEvent.TYPE_WINDOW_CONTENT_CHANGED);
+  }
+
```
removing this will remove the error message announcement?

fabriziobertoglio1987 commented on ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java:
```diff
@@ -333,7 +333,7 @@ private ReactTextUpdate getReactTextUpdate(
 
     return new ReactTextUpdate(
-        sb, mostRecentEventCount, false, 0, 0, 0, 0, Gravity.NO_GRAVITY, 0, 0, start, end);
+        sb, mostRecentEventCount, false, 0, 0, 0, 0, Gravity.NO_GRAVITY, 0, 0, start, end, null);
```
should this return the current value `editText.getError()`. It is triggered with updates from javascript

fabriziobertoglio1987 commented on ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java:
```diff
@@ -318,21 +318,35 @@ public void receiveCommand(
 
         if (!args.isNull(1)) {
           String text = args.getString(1);
+          String accessibilityErrorMessage = (String) reactEditText.getError();
```
does it return the error value even we set it with node.setError()?

fabriziobertoglio1987 commented:
<details><summary>Fabric - trigger the error onChangeText/summary>
<p>

https://user-images.githubusercontent.com/24992535/200121108-de546f2f-2238-44d9-a028-a69fe05183bc.mp4

</p>
</details>

fabriziobertoglio1987 commented:
<details><summary>Fabric Android - triggering the error with a button (outside of onChangeText callback)</summary>
<p>



https://user-images.githubusercontent.com/24992535/200121219-1189d00f-bb44-46e2-941b-43740b3792f5.mp4



</p>
</details>

fabriziobertoglio1987 commented:
<details><summary>Fabric Android - announce error after changing text value</summary>
<p>



https://user-images.githubusercontent.com/24992535/200121367-e4ce5324-ef2a-48d2-b391-c56d28d9ee94.mp4


</p>
</details>

fabriziobertoglio1987 commented:
<details><summary>Paper Android - setting accessibilityError onChangeText</summary>
<p>

https://user-images.githubusercontent.com/24992535/200121527-26a193cb-a85c-466d-8df1-7c23abea3a82.mp4




</p>
</details>

fabriziobertoglio1987 commented:
<details><summary>Fabric iOS - triggering accessibilityErrorMessage onChangeText</summary>
<p>


https://user-images.githubusercontent.com/24992535/200287750-44973c43-59e2-4268-8ee4-9916b53330f5.mp4



</p>
</details>

lunaleaps commented on Libraries/Text/TextInput/Multiline/RCTUITextView.h:
```diff
@@ -26,6 +26,8 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, assign, readonly) BOOL textWasPasted;
 @property (nonatomic, copy, nullable) NSString *placeholder;
 @property (nonatomic, strong, nullable) UIColor *placeholderColor;
+@property (nonatomic, readwrite, nullable) NSString *currentAccessibilityError;
```
@fabriziobertoglio1987 I'm getting a build error because this differs from Libraries/Text/TextInput/Singleline/RCTUITextField.h

Re
```
stderr: In file included from xplat/js/react-native-github/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:15:
xplat/js/react-native-github/Libraries/Text/TextInput/Singleline/RCTUITextField.h:32:54: error: 'retain (or strong)' attribute on property 'accessibilityErrorMessage' does not match the property inherited from 'RCTBackedTextInputViewProtocol' [-Werror,-Wproperty-attribute-mismatch]
[@property](https://www.internalfb.com/intern/profile/property) (nonatomic, readwrite, nullable) NSString *accessibilityErrorMessage;
```

Which property declaration should we use? cc @blavalla if you have thoughts

fabriziobertoglio1987 replied:
@lunaleaps The RCTTextInputComponentView.mm is the iOS Fabric Component for TextInput. Creates an instance _backedTextInput of type [BackedTextInputViewProtocol](https://github.com/facebook/react-native/blob/f0d0186ebad7312cdcd54b74d8d47e47bd4d741b/Libraries/Text/TextInput/RCTBackedTextInputViewProtocol.h#L20) using RCTUITextView or RCTUITextField constructors.
https://github.com/facebook/react-native/blob/f0d0186ebad7312cdcd54b74d8d47e47bd4d741b/React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm#L78

BackedTextInputViewProtocol, RCTUITextView and RCTUITextField headers should match (currently they don't, accessibilityErrorMessage does not exist in RCTBackedTextInputViewProtocol).
Sorry for the mistake. I'm rebasing, building, testing and pushing the commit soon. Thanks

<!--https://stackoverflow.com/questions/2255861/property-retain-assign-copy-nonatomic-in-objective-c-->

fabriziobertoglio1987 replied:
https://github.com/facebook/react-native/pull/33468/commits/9d1adecb7a58852f50ba43cc758f5e0d19b81e2d

https://user-images.githubusercontent.com/24992535/205132148-e7024469-1f9c-4c57-bfcf-476d21bbf5db.mp4

lunaleaps reviewed (changes requested):
Hey @fabriziobertoglio1987 I'm getting another build error now: 

```
stderr: xplat/js/react-native-github/Libraries/Text/TextInput/Multiline/RCTUITextView.m:16:17: error: auto property synthesis will not synthesize property 'accessibilityErrorMessage' declared in protocol 'RCTBackedTextInputViewProtocol' [-Werror,-Wobjc-protocol-property-synthesis]
@implementation RCTUITextView {
                ^
xplat/js/react-native-github/Libraries/Text/TextInput/RCTBackedTextInputViewProtocol.h:18:51: note: property declared here
@property (nonatomic, assign, nullable) NSString *accessibilityErrorMessage;
                                                  ^
xplat/js/react-native-github/Libraries/Text/TextInput/Multiline/RCTUITextView.m:324:1: note: add a '@synthesize' directive
@end
^
1 error generated.
    When running <objective-c preprocess_and_compile>.
    When building rule //xplat/js/react-native-github:RCTTextApple#compile-RCTUITextView.m.o92c2ff30,iphonesimulator-x86_64."
```

Do you know how to address?

fabriziobertoglio1987 commented:
Sorry @lunaleaps I had to rebase the branch 2 times. I missed some changes when reapplying the diff after the second rebase.

<details><summary>The warning in RCTUITextView</summary>
<p>

<img width="1920" alt="Screenshot 2022-12-05 at 9 58 55 PM" src="https://user-images.githubusercontent.com/24992535/205741525-af675e23-cd59-46d6-964a-f3885fc4456e.png">

</p>
</details>

<details><summary>Adding the missing header accessibilityErrorMessage in RCTUITextView</summary>
<p>

<img width="1920" alt="Screenshot 2022-12-05 at 9 59 25 PM" src="https://user-images.githubusercontent.com/24992535/205741611-8a75b6e6-7585-481d-885c-bb79ed7216f9.png">

</p>
</details>

<details><summary>The header accessibilityErrorMessage was already present in RCTBackedTextInputViewProtocol</summary>
<p>


<img width="1920" alt="Screenshot 2022-12-05 at 10 05 36 PM" src="https://user-images.githubusercontent.com/24992535/205742616-350279a9-06e7-4669-bc16-1fff033923ac.png">


</p>
</details>

<details><summary>The warning disappears after adding header in RCTUITextView</summary>
<p>

<img width="1920" alt="Screenshot 2022-12-05 at 10 02 51 PM" src="https://user-images.githubusercontent.com/24992535/205742140-309968e0-9133-4c32-b7d8-d003fcf7ffd7.png">

</p>
</details>

https://github.com/facebook/react-native/pull/33468/commits/ca6d8119cdbd232503312a5b6159bb9b7cbe03bf

fabriziobertoglio1987 commented on Libraries/Text/TextInput/RCTBackedTextInputViewProtocol.h:
```diff
@@ -15,6 +15,9 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, copy, nullable) NSAttributedString *attributedText;
+@property (nonatomic, assign, nullable) NSString *accessibilityErrorMessage;
+@property (nonatomic, readwrite, nullable) NSString *currentAccessibilityError;
+@property (nonatomic, readwrite, nullable) NSString *previousAccessibilityError;
```
to be removed

lunaleaps commented:
Great, re-importing and verifying on our internal CI. I think if all is good, we'll probably need to land this after the holiday break due to code freeze

lunaleaps reviewed (changes requested):
See comment on build error

fabriziobertoglio1987 commented:
@lunaleaps https://github.com/facebook/react-native/pull/33468#issuecomment-1387675981

I prepared [improvements](https://github.com/facebook/react-native/compare/main...fabriziobertoglio1987:react-native:text-input-errors-improvements?expand=1#diff-589692b2662a9e15ae1a961b98289ba4c4b0231d448e985066a29c68f617db5fR145-R644) to the iOS API, and they are the solution for this issue.
I want to publish a separate Pull Request for the iOS functionalities. I already tested the new [branch](https://github.com/facebook/react-native/compare/main...fabriziobertoglio1987:react-native:text-input-errors-improvements?expand=1#diff-589692b2662a9e15ae1a961b98289ba4c4b0231d448e985066a29c68f617db5fR145-R644) with the improvements. 
The only work required is deleting iOS logic from PR #33468 and publishing a separate Pull Request for the iOS functionalities. The improvements already included in the new branch are:

1) removed logic from `setAccessibilityErrorMessage` in `Singleline/RCTUITextField` and `Multiline/RCTUITextView`. The method `setAccessibilityErrorMessage` should only update `accessibilityErrorMessage`. The method includes some irrelevant logic (updating the accessibilityValue and triggering VoiceOver manual announcement).
https://github.com/fabriziobertoglio1987/react-native/commit/fa6de6ab5335fe56b8d11a62cd16ad901d3ca093#diff-34e52755149f4f49b5b68b5152a1e81df1671422e38dca7eab1751c86131aa31L102-L118

<details><summary>CLICK TO OPEN SOURCECODE</summary>
<p>


<img width="1482" alt="Screenshot 2023-01-19 at 12 30 30 PM" src="https://user-images.githubusercontent.com/24992535/213431433-73c5696d-80e4-4b0b-9290-c5c9f1fa7569.png">

</p>
</details>

2) moved the previous logic to `RCTTextInputComponentView.mm` method `updateProps`
https://github.com/fabriziobertoglio1987/react-native/commit/fa6de6ab5335fe56b8d11a62cd16ad901d3ca093#diff-589692b2662a9e15ae1a961b98289ba4c4b0231d448e985066a29c68f617db5fL145-R637 

<details><summary>CLICK TO OPEN SOURCECODE</summary>
<p>

<img width="1482" alt="Screenshot 2023-01-19 at 12 31 19 PM" src="https://user-images.githubusercontent.com/24992535/213431601-2fb1ed25-9002-4340-bc06-5bf8e72d6599.png">

</p>
</details>

3) Optional - further improve the iOS solution.

Relevant: https://github.com/facebook/react-native/pull/33468#discussion_r1037410975
Let me know if you are happy with this solution otherwise I'm happy to follow your requirements https://github.com/facebook/react-native/pull/33468#issuecomment-1387675981

lunaleaps commented:
Hi @fabriziobertoglio1987,

Do you mean you want to start a new PR for these changes and abandon this one? Or do you mean you're splitting this PR up into 2 PRs. Should I continue trying to merge this PR?

fabriziobertoglio1987 commented:
> Or do you mean you're splitting this PR up into 2 PRs. Should I continue trying to merge this PR?

@lunaleaps https://github.com/facebook/react-native/pull/33468#issuecomment-1387675981 I can remove the iOS functionalities from this PR, and you can merge the Android functionalities. I can publish an improved version of the iOS functionalities in a separate PR https://github.com/facebook/react-native/pull/33468#issuecomment-1396841801, if you desire it or I can apply the changes to fix the error reported here https://github.com/facebook/react-native/pull/33468#issuecomment-1387675981 

I'm happy to do as you like. 🙏

lunaleaps commented:
@fabriziobertoglio1987 Either is fine for me. Perhaps splitting the platforms up (if it's possible) would be easier.

fabriziobertoglio1987 commented:
https://user-images.githubusercontent.com/24992535/213680513-0c3029da-ed51-4beb-a0d6-399195fbbc5b.mp4


https://user-images.githubusercontent.com/24992535/213680534-bcfbd1e1-5729-438d-bf2b-4fce4bbba01f.mp4

fabriziobertoglio1987 commented:
https://user-images.githubusercontent.com/24992535/213683985-4667bbf5-bf11-45ca-a1a9-5d8c71a2f617.mp4

fabriziobertoglio1987 commented:
@lunaleaps I moved the iOS functionalities to https://github.com/facebook/react-native/pull/35908. I will move https://github.com/facebook/react-native/pull/35908 to ready to review when this PR is merged. Thanks a lot.
